sdk/lib/rtl/rangelist.c

```diff
@@ -13,546 +13,1556 @@
-            RangeList->Stamp++;
-            return STATUS_SUCCESS;
-        }
+    *OutRange = (PRTL_RANGE)Iterator->Current;
 
-        Entry = Entry->Flink;
-    }
+    DPRINT("RtlGetLastRange: *OutRange %p, [%I64X-%I64X]\n", Iterator->Current, (*OutRange)->Start, (*OutRange)->End);
 
-    return STATUS_RANGE_NOT_FOUND;
+    return STATUS_SUCCESS;
 }
 
-
-/**********************************************************************
- * NAME							EXPORTED
- * 	RtlFindRange
+/**
+ * @name RtlFindRange
  *
- * DESCRIPTION
- *	Searches for an unused range.
+ * Searches for an unused range.
  *
- * ARGUMENTS
- *	RangeList		Pointer to the range list.
- *	Minimum
- *	Maximum
- *	Length
- *	Alignment
- *	Flags
- *	AttributeAvailableMask
- *	Context
- *	Callback
- *	Start
+ * @param RangeList
+ * Pointer to the range list.
  *
- * RETURN VALUE
- *	Status
+ * @param Minimum
+ * Minimal value for search.
  *
- * TODO
- *	Support shared ranges and callback.
+ * @param Maximum
+ * Maximumal value for search.
  *
- * @implemented
+ * @param Length
+ * Length of the range.
+ *
+ * @param Alignment
+ * Alignment value.
+ *
+ * @param Flags
+ * Flags describing the control the behavior of lists.
+ *
+ * @param AttributeAvailableMask
+ * Mask for Attributes from range entry.
+ *
+ * @param Context
+ * Context for Callback.
+ *
+ * @param Callback
+ * Pointer to a callback to call in case of a conflict.
+ *
+ * @param OutStart
+ * Pointer where the found value for the start of the range will be stored.
+ *
+ * @return
+ * Returns NTSTATUS.
  */
+CODE_SEG("PAGE")
+NTSYSAPI
 NTSTATUS
 NTAPI
-RtlFindRange(IN PRTL_RANGE_LIST RangeList,
-             IN ULONGLONG Minimum,
-             IN ULONGLONG Maximum,
-             IN ULONG Length,
-             IN ULONG Alignment,
-             IN ULONG Flags,
-             IN UCHAR AttributeAvailableMask,
-             IN PVOID Context OPTIONAL,
-             IN PRTL_CONFLICT_RANGE_CALLBACK Callback OPTIONAL,
-             OUT PULONGLONG Start)
+RtlFindRange(
+    _In_ PRTL_RANGE_LIST RangeList,
+    _In_ ULONGLONG Minimum,
+    _In_ ULONGLONG Maximum,
+    _In_ ULONG Length,
+    _In_ ULONG Alignment,
+    _In_ ULONG Flags,
+    _In_ UCHAR AttributeAvailableMask,
+    _In_ PVOID Context OPTIONAL,
+    _In_ PRTL_CONFLICT_RANGE_CALLBACK Callback OPTIONAL,
+    _Out_ PULONGLONG OutStart)
 {
-    PRTL_RANGE_ENTRY CurrentEntry;
-    PRTL_RANGE_ENTRY NextEntry;
-    PLIST_ENTRY Entry;
-    ULONGLONG RangeMin;
-    ULONGLONG RangeMax;
-
-    if (Alignment == 0 || Length == 0)
+    RTL_RANGE_LIST_ITERATOR Iterator;
+    PRTL_RANGE TmpRange;
+    ULONGLONG Start;
+    ULONGLONG End;
+
+    PAGED_CODE_RTL();
+    DPRINT("RtlFindRange: [%X] %p, %X, [%I64X-%I64X], %X, %X\n", Flags, RangeList, RangeList->Count, Minimum, Maximum, Length, Alignment);
+
+    ASSERT(RangeList);
+    ASSERT(OutStart);
+    ASSERT(Alignment > 0);
+    ASSERT(Length > 0);
+
+    Start = Maximum - (Length - 1) - ((Maximum - (Length - 1)) % Alignment);
```

Comment:
Please introduce a human-readable `ROUND_DOWN`-type macro for ULONGLONGs to be used here (and possibly in at least one other place). This should correspond to:
```suggestion
    Start = ROUND_DOWN(Maximum - (Length - 1), Alignment);
```

Reply:
Maybe
#define ROUND64_DOWN_BY(size,align)  (((ULONGLONG)(size)) & (~((ULONGLONG)align - 1)))
#define ROUND64_DOWN_BY(size,align)  ((ULONGLONG)size - ((ULONGLONG)size % (ULONGLONG)align))